### PR TITLE
Add Chart Builder

### PR DESCRIPTION
**Pull request by cristineguadelupe**
Initial implementation of Chart Builder

In this initial stage, we have a "UI to chart" where the goal is to allow the creation of the most used charts from a simple interface.

https://user-images.githubusercontent.com/5660941/158725512-88fbce34-ec3b-459a-bb71-ce2f9491a3c3.mp4

CleanShot 2022-03-16 at 23 40 33@2x

CleanShot 2022-03-16 at 23 40 46@2x

CleanShot 2022-03-16 at 23 13 06@2x

### CODE REVIEW

File: lib/kino/smart_cell/chart_builder.ex
Comment (polvalente) [inline lib/kino/smart_cell/chart_builder.ex:110]:
```suggestion
      Map.new(attrs, fn {k, v} -> {String.to_atom(k), v} end)
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (polvalente) [inline lib/kino/smart_cell/chart_builder.ex:113]:
```suggestion
      %{field: :new, function_name: :new, module: module, args: [[width: attrs.width, height: attrs.height]]},
```

I was thinking about this data structure, and now it seems to me that abandoning the outer keyword in favor of a list of maps could be better.

Some matches below would also need to change, but it makes for a cleaner code overall. `[root | nodes] = ...` for instance

File: lib/kino/smart_cell/chart_builder.ex
Comment (polvalente) [inline lib/kino/smart_cell/chart_builder.ex:124]:
```suggestion
    |> Enum.map(&clean_node/1)
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (polvalente) [inline lib/kino/smart_cell/chart_builder.ex:163]:
```suggestion
      |> Enum.filter(&is_list/1)
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (polvalente) [inline lib/kino/smart_cell/chart_builder.ex:170]:
```suggestion
    args = Enum.reject(args, & is_list(&1) or is_nil(&1))
```

File: lib/assets/chart_builder/main.js
Comment (polvalente) [inline lib/assets/chart_builder/main.js:112]:
```suggestion
  ctx.handleEvent("update", ({ fields }) => setValues(fields));
```

can't these be shortened to one liners like so?

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:28]: From the (unpublished) docs:

> **Important:** remember that data sent between processes is copied,
> so avoid sending large data structures. In particular, when looking
> at variables, instead of sending their values, extract and send
> only the relevant metadata.

So basically we want to move the work from `handle_info` here and then `send(pid, {:scan_binding_result, data_options, vl_alias})`, this way we send the minimal amount of data between processes :)

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:136]: There's a catch! Let's say we have a variable with data and the chart builder below using this variable. Now when we re-open the notebook (or restart the runtime and install the deps), the smart cell would start, but it would receive an empty binding, until the above code is executed. In such case, we want to preserve the previously selected variable name (that we persist via `attrs`) until the user changes it, regardless if it's currently available. This is important when there are more options, because we don't want to change it implicitly. It's analogous to the connection variable in #105, here's an example:

https://user-images.githubusercontent.com/17034772/158814743-f5fe9837-b701-4900-b899-5627484f8271.mp4

We can handle it the same way, in particular see the `function renderConnectionSelect` function.

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:82]: It may not be the only alias, so we should do `Enum.find_value` instead :)

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:97]: Once #105 is merged

```suggestion
    attrs |> to_quoted() |> Kino.Utils.Code.quoted_to_string()
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:7]: Nit: it's better to keep those as strings and convert to atoms only when generating the code. This way we keep the same type as serialized/deserialized in `attrs` and there's gonna be less surprises :)

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:99]:
```suggestion
  const channelFields = ["x_axis", "y_axis", "color"];
  const typeFields = ["x_axis_type", "y_axis_type", "color_type"];
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:18]: We want to order from the most common type, but I'm not sure how to establish this. @josevalim any opinion?

```suggestion
                <option value="point">Point</option>
                <option value="line">Line</option>
                <option value="bar">Bar</option>
                <option value="area">Area</option>
```

Comment (josevalim) [reply lib/assets/chart_builder/main.js:18]: No idea. Why not alphabetically? :)

Comment (jonatanklosko) [reply lib/assets/chart_builder/main.js:18]: In this particular case alphabetically doesn't seem right, because `Area` is probably the rarest of these four? :smile:

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:157]: I think it can be something like this:

```suggestion
    args = if function == :encode_field, do: List.insert_at(args, 0, field), else: args
    
    quote do
      unquote(acc) |> unquote(module).unquote(function)(unquote_splicing(args))
    end
```

We could likely simplify other places with `quote` too, but it's gonna be easier to refactor when we have some tests for the code generation :) You can test the `to_source` function directly as we do [here](https://github.com/livebook-dev/kino/blob/ead0a6f42727eeec95c450b3bb5632c6ea9133e2/test/kino/smart_cell/sql_test.exs#L63-L173).

File: lib/assets/chart_builder/main.js
Comment (josevalim) [inline lib/assets/chart_builder/main.js:220]:
```suggestion
The dataset needs to be a map of series, for example:

      my_data = %{
        a: [89, 124, 09, 67, 45],
        b: [12, 45, 67, 83, 32]
      }

Or using Explorer:

      iris = Explorer.Datasets.iris() |> Explorer.DataFrame.to_map()

Chart builder is available for all datasets above.`;
```

Comment (jonatanklosko) [reply lib/assets/chart_builder/main.js:220]: > Chart builder is available for all datasets above

I'm wondering if it makes sense to say this, given it's the core principle in Livebook in general.

File: lib/assets/chart_builder/main.js
Comment (cristineguadelupe) [reply lib/assets/chart_builder/main.js:18]: I feel that Bar and Line are the most common/familiar (especially for beginners) so my suggestion is Bar, Line, Point, Area. Wdy?

File: lib/assets/chart_builder/main.js
Comment (cristineguadelupe) [reply lib/assets/chart_builder/main.js:236]: This is cool but I need more help here! I ended up with more code for the elements where I need to append or prepend options. Or with two different ways to create the options depending on the rules.

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:105]:
```suggestion
    if (channelFields.includes(name)) { maybeReset(name, value) };
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:190]:
```suggestion
      selectElem.disabled = (elem === "color_type");
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [reply lib/assets/chart_builder/main.js:236]: I'd say it's enough to make the select dimmed, rather than individual options (that's what we do for the SQL cell). Then you could conditionally prepend an option to the list and render all of them. That said, I'm ok with either approach, and creating elements directly is fine, so your call :)

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:168]: The data info message no longer shows, this should probably be

```suggestion
    updateDataInfoBox(options[selected]);
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:192]: I think you can simplify if you make `:opts` a separate key in the node, since we know what any options should go as the last argument :)

**Review by jonatanklosko — COMMENTED** [review]
Awesome, a couple more comments from me :)

File: lib/kino/smart_cell/chart_builder.ex
Comment (josevalim) [inline lib/kino/smart_cell/chart_builder.ex:99]:
```suggestion
    case List.keyfind(aliases, VegaLite, 1) do
      {alias, _} -> alias
      nil -> nil
    end
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (josevalim) [reply lib/kino/smart_cell/chart_builder.ex:192]: I would propose another approach instead. Instead of having the node do this:

        args: [attrs.color, [type: attrs.color_type]]

I think you should have something like this:

        args: maybe_arg_type(attrs.color, attrs.color_type)

Which builds the correct argument in the first place (or set it to `nil` so it is discarded later).

File: lib/kino/smart_cell/chart_builder.ex
Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex:22]: It would be okay by me, but I honestly don't see the benefits. We already have consistency since we have empyt strings for JS and `nil` for Elixir, and the conversion happens only on `to_source`. I also think it would lead us to have comparisons with  `=== "nil"` on JS's side which doesn't hurt but seems a bit odd... Unless I'm missing something here 🤔

File: lib/assets/chart_builder/main.js
Comment (cristineguadelupe) [reply lib/assets/chart_builder/main.js:238]: I'll keep it as it is for now! :)

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [reply lib/kino/smart_cell/chart_builder.ex:22]: Ah, I wrote that before I realised that `convert_field` doesn't happen on change, but before code generation.

I'm mostly wondering about the types we should have in `attrs` (because that's what we persist and it's hard to change later on). It may be convenient to just keep the values coming from JS, which are pretty much always strings, but I think keeping the parsed values is cleaner. So for fields that are selects either value or `nil`, for numbers either integer or `nil`, for text fields always string. In other words, I'd lean towards parsing the data on change, though I don't have a strong reason other than it _feels_ better to me and I believe that's what we do in the other cells :)

> I also think it would lead us to have comparisons with `=== "nil"`

`nil` becomes `null` on JS side :)

Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex:22]: Done 🐈

File: lib/kino/smart_cell/chart_builder.ex
Comment (josevalim) [inline lib/kino/smart_cell/chart_builder.ex:138]: You could do `build_arg_root(height, width)` here and simplify the build_root later on. :) similar to what we did with `arg_type`!

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:41]: When we generate the code with alias and then open the notebook, as soon as the cell starts (the first cell gets evaluated) we would generate a new code without the alias, and then update once the alias is created. I'd say it's fine, @josevalim wdyt?

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:71]: Note that we compare the value below, so we want to use the parsed one for that.

```suggestion
    value = parse_value(field, value)
    ctx = update(ctx, :fields, &Map.put(&1, field, value))
```

(in the db connection cell we have `to_updates`, because changing database type also updates the port, here it's always a single update)

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:73]: Doesn't the client have all options from the `"set_available_data"` event?

File: lib/kino/smart_cell/chart_builder.ex
Comment (josevalim) [reply lib/kino/smart_cell/chart_builder.ex:41]: @jonatanklosko fine by me.

File: lib/kino/smart_cell/chart_builder.ex
Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex]: Done!

**Comment by cristineguadelupe** [comment]
> Should we rather the infobox generic and build the error message from Elixir? This way we write less JavaScript... something like this:
> 
> ```
>   ctx.handleEvent("error", ({ message }) => {
>     updateInfoBox(message);
> ```

Since we are moving more things to JS I’m not sure if that’s make sense for now

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:226]:
```suggestion
    if (value) {
      elem.disabled = false;
    } else {
      resetField(elem)
    }
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:225]:
```suggestion
  function maybeResetChannelType(name, value) {
    const typeElem = ctx.root.querySelector(`[name="${name}_type"]`);
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:145]: ?

```suggestion
    const axisOptions = options[selectedData];
```

File: lib/assets/chart_builder/main.js
Comment (jonatanklosko) [inline lib/assets/chart_builder/main.js:193]: We are sending `"update_field"` in the loop above, and then in `setTypes` in a loop as well. We shouldn't be spamming the server so much.

This bulk change is a result of selecting the dataframe, to set the new defaults we can either:

1. Compute the new values on the client and send as multi-field update, then broadcast all those fields
2. Send just the new dataframe name and determine the new values on the server, then broadcast all those fields

I think 2. is more direct and will end up better :)

**Review by jonatanklosko — COMMENTED** [review]
Looks great, just one more important detail and we should be good :)

**Review by josevalim — APPROVED** [review]
All good from my side. The JS is up to Jonatan :D

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:89]: Would it make more sense to default x and y to different columns?

```suggestion
    {x_field, y_field} =
      case ctx.assigns.options[df] do
        [] -> {nil, nil}
        [key] -> {Atom.to_string(key), Atom.to_string(key)}
        [key1, key2 | _] -> {Atom.to_string(key1), Atom.to_string(key2)}
      end
```

Also, we could convert the column names to strings upfront in `scan_binding`, since we don't need them as atoms.

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:66]: Shouldn't we reset other fields in this case? We can match in `update_data_options` on empty value, and return all empty fields :)

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:86]: `update_` suggests this function changes `ctx` or has a side effect, maybe `updates_for_data_variable`?

**Review by jonatanklosko — COMMENTED** [review]
Final nitpicks and ship it! :)

File: lib/kino/smart_cell/chart_builder.ex
Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex:70]: Nops! Here we have a specific case. `data_variable` will be `""` if and only if the ChartBuilder has never been populated before. So we have nothing to reset and we don't need to broadcast.

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:43]:
```suggestion
          do: {Atom.to_string(key), val |> Map.keys() |> Enum.map(&to_string/1)}
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex:70]: This happens in two scenarios: when a ChartBuilder is started without variables and when reconnection with an empty (or fresh) ChartBuilder occurs.

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [inline lib/kino/smart_cell/chart_builder.ex:94]: Actually the map keys may as well be strings or numbers, so we probably want `to_string` instead, but we can as well do it when scanning too.

```suggestion
        [key] -> {key, key}
        [key1, key2 | _] -> {key1, key2}
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [reply lib/kino/smart_cell/chart_builder.ex:70]: What I meant is that if it's the JS side triggering and we don't want it, we can always add a check like

```js
if (name === "data_variable" && value === "") { return }
```

File: lib/kino/smart_cell/chart_builder.ex
Comment (jonatanklosko) [reply lib/kino/smart_cell/chart_builder.ex:43]: Ah, in fact the map may have keys that we cannot convert to strings, we should filter only those that have atom/string keys then.

(in the future we will have a separate abstraction for tabular data that will be more precise and flexible, so for now we just don't want to crash ^^)

Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex]: `data = %{3 => [1, 2, 3], 4 => [1, 2, 3]}` this is working! Are there other cases?

Comment (jonatanklosko) [reply lib/kino/smart_cell/chart_builder.ex]: The problem is with terms that don't support `to_string`, like this:

```
data = %{self() => [1, 2, 3]}
```

Comment (cristineguadelupe) [reply lib/kino/smart_cell/chart_builder.ex]: I see! I tried to solve it for now. Thanks!

**Review by jonatanklosko — APPROVED** [review]
Fantastic :cat:

**Pull request merged**
